server/lib/job-queue/handlers/video-channels-sync.ts

```diff
@@ -0,0 +1,266 @@
+  } catch (ex) {
+    logger.error(`Failed to import channel ${videoChannel.name}: ${ex.stack}`)
+  }
+}
+
+type SynchronizeChannelOptions = {
+  youtubeDL: YoutubeDLCLI
+  secondsToWait: number
+  lastVideosCount?: number
+  after?: string
+}
+async function synchronizeChannel (channel: VideoChannelModel, externalChannelUrl: string, { youtubeDL, secondsToWait, lastVideosCount, after }: SynchronizeChannelOptions): Promise<ChannelSyncInfo> {
+  const result: ChannelSyncInfo = {
+    total: VIDEO_CHANNEL_MAX_SYNC,
+    errors: 0,
+    successes: 0,
+    alreadyImported: 0
+  }
+  const user = await UserModel.loadByChannelActorId(channel.actorId)
+  const additionalYoutubeDLArgs = [ '--skip-download', '--playlist-reverse' ]
+  if (lastVideosCount) {
+    additionalYoutubeDLArgs.push('--playlist-end', VIDEO_CHANNEL_MAX_SYNC.toString());
+  }
+  const channelInfo = await youtubeDL.getInfo({
+    url: externalChannelUrl,
+    format: YoutubeDLCLI.getYoutubeDLVideoFormat([]),
+    processOptions,
+    additionalYoutubeDLArgs,
+  })
+  const targetUrls: string[] = (await Promise.all(
+    channelInfo.map(async video => {
+      if (after && video['upload_date'] <= after) {
+        return []
+      }
+      return video['webpage_url']
+    })
+  )).flat()
+
+  await waitSecs(secondsToWait)
+
+  for (const targetUrl of targetUrls) {
+    try {
+      // TODO retry pour l'import d'une chaîne ?
+      if (!await VideoImportModel.urlAlreadyImported(user.id, targetUrl)) {
+        await addYoutubeDLImport({
+          user,
+          channel,
+          targetUrl
+        })
+        result.successes += 1
+      } else {
+        result.alreadyImported += 1
+      }
+    } catch (ex) {
+      result.errors += 1
+      logger.error(`An error occured while importing ${targetUrl}: ${ex.stack}`)
+    }
+    await waitSecs(secondsToWait)
+  }
+  return result
+}
+
+async function buildVideo (channelId: number, targetUrl: string, importData: YoutubeDLInfo): Promise<MVideoThumbnail> {
+  let videoData = {
+    name: importData.name ?? 'Unknown name',
+    remote: false,
+    category: importData.category,
+    licence: importData.licence ?? CONFIG.DEFAULTS.PUBLISH.LICENCE,
+    language: importData.language,
+    commentsEnabled: CONFIG.DEFAULTS.PUBLISH.COMMENTS_ENABLED,
+    downloadEnabled: CONFIG.DEFAULTS.PUBLISH.DOWNLOAD_ENABLED,
+    waitTranscoding: false,
+    state: VideoState.TO_IMPORT,
+    nsfw: importData.nsfw ?? false,
+    description: importData.description,
+    support: null,
+    privacy: VideoPrivacy.PUBLIC,
+    duration: 0, // duration will be set by the import job
+    channelId: channelId,
+    originallyPublishedAt: importData.originallyPublishedAt
+  }
+
+  videoData = await Hooks.wrapObject(
+    videoData,
+    'filter:api.video.import-url.video-attribute.result'
+  )
+
+  const video = new VideoModel(videoData)
+  video.url = getLocalVideoActivityPubUrl(video)
+
+  return video
+}
+
+async function addYoutubeDLImport (parameters: {
```

Comment:
Could you try to refactor this function? It shares a lot of code with the `addYoutubeDLImport` function form the controller

Reply:
I think I should implement test before this refactoring. Do you mind if I do that in this PR? (I would understand as it is a bit long enough)

Reply:
I'd rather have the tests included in this PR. But I can work on them with you if you want

Reply:
I start to work on it :)

Reply:
Done a1f86312d8fa1178b1df9b41ae2bfc7afc88a481